File: 977_sorted_squares.py
from typing import List


class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        slow, fast, result = 0 , len(nums)-1 , len(nums)-1
        res = [float('inf')]*len(nums)
        while slow <= fast:
            if((nums[slow] * nums[slow]) > (nums[fast] * nums[fast])):
               res[result] = nums[slow] * nums[slow]
               slow+=1
            else:
                res[result] = nums[fast] * nums[fast]
                fast-=1

            result+=1
        return res

File: 977_sorted_squares.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* sortedSquares(int* nums, int numsSize, int* returnSize){
    *returnSize = numsSize;
    int slow = 0;
    int fast = numsSize - 1;
    int* res = (int*)malloc(sizeof(int) * numsSize);
    int index;
    for(index = numsSize - 1;index >= 0;index--)
    {
        int fastSqr = nums[fast] * nums[fast];
        int slowSqr = nums[slow] * nums[slow];
        if(slowSqr > fastSqr)
        {
            res[index] = slowSqr;
            slow++;
        }
        else
        {
            res[index] = fastSqr;
            fast--;
        }
        
    }
    return res;
}